File: native-ios/GoWithUs/Services/NotificationService.swift
import Foundation

class NotificationService {
    static let shared = NotificationService()
    
    private init() {}
    
    // MARK: - Get Notifications
    func getNotifications() async throws -> [AppNotification] {
        return try await APIService.shared.request(
            endpoint: "/notifications",
            method: .get
        )
    }
    
    // MARK: - Mark as Read
    func markAsRead(id: String) async throws {
        let _: EmptyResponse = try await APIService.shared.request(
            endpoint: "/notifications/\(id)/read",
            method: .put
        )
    }
    
    // MARK: - Create Notification (Admin only)
    func createNotification(title: String, message: String, type: String, targetId: String?) async throws -> AppNotification {
        let request = CreateNotificationRequest(
            title: title,
            message: message,
            type: type,
            targetId: targetId
        )
        
        return try await APIService.shared.request(
            endpoint: "/notifications",
            method: .post,
            body: request
        )
    }
    
    // MARK: - Get Unread Count
    func getUnreadCount() async throws -> Int {
        struct UnreadCountResponse: Decodable {
            let count: Int
        }
        
        let response: UnreadCountResponse = try await APIService.shared.request(
            endpoint: "/notifications/unread-count",
            method: .get
        )
        
        return response.count
    }
    
    // MARK: - Delete Notification
    func deleteNotification(id: String) async throws {
        let _: EmptyResponse = try await APIService.shared.request(
            endpoint: "/notifications/\(id)",
            method: .delete
        )
    }
    
    // MARK: - Clear All Notifications
    func clearAllNotifications() async throws {
        let _: EmptyResponse = try await APIService.shared.request(
            endpoint: "/notifications/clear-all",
            method: .delete
        )
    }
    // MARK: - Register Device Token
    func registerDeviceToken(token: String) async throws {
        struct TokenRequest: Encodable {
            let token: String
        }
        
        let request = TokenRequest(token: token)
        
        let _: EmptyResponse = try await APIService.shared.request(
            endpoint: "/users/device-token",
            method: .post,
            body: request
        )
    }
}

// Empty response for endpoints that don't return data
private struct EmptyResponse: Decodable {}

File: web-frontend/src/services/notificationService.ts
import { authFetch } from './api';

const API_BASE_URL = import.meta.env.VITE_API_URL || '/api';

export interface Notification {
    id: string;
    title: string;
    message: string;
    type: 'alert' | 'trip' | 'system';
    targetId?: string;
    createdAt: string;
    isRead: boolean;
}

export interface CreateNotificationRequest {
    title: string;
    message: string;
    type: string;
    targetId?: string;
}

export const notificationService = {
    // Get notifications with optional 'all' flag for admins
    async getNotifications(all: boolean = false): Promise<Notification[]> {
        const response = await authFetch(`${API_BASE_URL}/notifications${all ? '?all=true' : ''}`);
        return response;
    },

    // Get unread notification count
    async getUnreadCount(): Promise<number> {
        const data = await authFetch(`${API_BASE_URL}/notifications/unread-count`);
        return data.count;
    },

    // Mark notification as read
    async markAsRead(id: string): Promise<void> {
        await authFetch(`${API_BASE_URL}/notifications/${id}/read`, {
            method: 'PUT',
        });
    },

    // Create notification (admin only)
    async createNotification(data: CreateNotificationRequest): Promise<Notification> {
        return authFetch(`${API_BASE_URL}/notifications`, {
            method: 'POST',
            body: JSON.stringify(data),
        });
    },

    // Delete single notification
    async deleteNotification(id: string): Promise<void> {
        await authFetch(`${API_BASE_URL}/notifications/${id}`, {
            method: 'DELETE',
        });
    },

    // Clear notifications (can be system-wide for admins)
    async clearAllNotifications(all: boolean = false): Promise<void> {
        await authFetch(`${API_BASE_URL}/notifications/clear-all${all ? '?scope=all' : ''}`, {
            method: 'DELETE',
        });
    },
};
